src/proof.rs

```diff
@@ -0,0 +1,1524 @@
+    // if !point.is_in_correct_subgroup_assuming_on_curve() {...}
+
+    Ok(point)
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct ZKProof {
+    // Commitments to wire polynomials
+    pub w1: G1ProofPoint,
+    pub w2: G1ProofPoint,
+    pub w3: G1ProofPoint,
+    pub w4: G1ProofPoint,
+    // Commitments to logup witness polynomials
+    pub lookup_read_counts: G1ProofPoint,
+    pub lookup_read_tags: G1ProofPoint,
+    pub lookup_inverses: G1ProofPoint,
+    // Commitment to grand permutation polynomial
+    pub z_perm: G1ProofPoint,
+    pub libra_commitments: [G1ProofPoint; LIBRA_COMMITMENTS],
+    // Sumcheck
+    pub libra_sum: Fr,
+    pub sumcheck_univariates: [[Fr; ZK_BATCHED_RELATION_PARTIAL_LENGTH]; CONST_PROOF_SIZE_LOG_N],
+    pub sumcheck_evaluations: [Fr; NUMBER_OF_ENTITIES],
+    pub libra_evaluation: Fr,
+    // ZK
+    pub gemini_masking_poly: G1ProofPoint,
+    pub gemini_masking_eval: Fr,
+    // Shplemini
+    pub gemini_fold_comms: [G1ProofPoint; CONST_PROOF_SIZE_LOG_N - 1],
+    pub gemini_a_evaluations: [Fr; CONST_PROOF_SIZE_LOG_N],
+    pub libra_poly_evals: [Fr; LIBRA_EVALUATIONS],
+    pub shplonk_q: G1ProofPoint,
+    pub kzg_quotient: G1ProofPoint,
+}
+
+impl TryFrom<&[u8]> for ZKProof {
+    type Error = ProofError;
+
+    fn try_from(proof_bytes: &[u8]) -> Result<Self, Self::Error> {
+        if proof_bytes.len() != ZK_PROOF_SIZE {
+            return Err(ProofError::IncorrectBufferSize {
+                expected_size: ZK_PROOF_SIZE,
+                actual_size: proof_bytes.len(),
+            });
+        }
+
+        let mut offset = 0;
+
+        // Commitments
+        let w1 = read_g1_proof_point(proof_bytes, &mut offset)?;
+        let w2 = read_g1_proof_point(proof_bytes, &mut offset)?;
+        let w3 = read_g1_proof_point(proof_bytes, &mut offset)?;
+
+        // Lookup / Permutation Helper Commitments
+        let lookup_read_counts = read_g1_proof_point(proof_bytes, &mut offset)?;
+        let lookup_read_tags = read_g1_proof_point(proof_bytes, &mut offset)?;
+        let w4 = read_g1_proof_point(proof_bytes, &mut offset)?;
+        let lookup_inverses = read_g1_proof_point(proof_bytes, &mut offset)?;
+        let z_perm = read_g1_proof_point(proof_bytes, &mut offset)?;
+
+        let mut libra_commitments = [G1ProofPoint::default(); LIBRA_COMMITMENTS];
+
+        libra_commitments[0] = read_g1_proof_point(proof_bytes, &mut offset)?;
+
+        let libra_sum = read_fr(proof_bytes, &mut offset)?;
+
+        // Sumcheck univariates
+        let mut sumcheck_univariates =
+            [[Fr::ZERO; ZK_BATCHED_RELATION_PARTIAL_LENGTH]; CONST_PROOF_SIZE_LOG_N];
+
+        for sumcheck_univariate in sumcheck_univariates.iter_mut().take(CONST_PROOF_SIZE_LOG_N) {
+            for su in sumcheck_univariate
+                .iter_mut()
+                .take(ZK_BATCHED_RELATION_PARTIAL_LENGTH)
+            {
+                *su = read_fr(proof_bytes, &mut offset)?;
+            }
+        }
```

Comment:
The `take` doesn't do anything here (same in the non-zk proof variant)
```suggestion
        for sumcheck_univariate in sumcheck_univariates.iter_mut() {
            for su in sumcheck_univariate.iter_mut() {
                *su = read_fr(proof_bytes, &mut offset)?;
            }
        }
```